jerry-core/ecma/operations/ecma-objects.c

```diff
@@ -2096,621 +2097,426 @@ ecma_op_object_get_enumerable_property_names (ecma_object_t *obj_p, /**< routine
 
-        prop_iter_cp = prop_iter_p->next_property_cp;
-      }
-    }
+  jmem_cpointer_t obj_cp;
+  ECMA_SET_NON_NULL_POINTER (obj_cp, obj_p);
 
-    uint32_t all_properties_count = array_index_named_properties_count + string_named_properties_count;
+  while (true)
+  {
+    ecma_collection_t *keys = ecma_op_object_own_property_keys (obj_p);
 
 #if ENABLED (JERRY_ESNEXT)
-    all_properties_count += symbol_named_properties_count;
+    if (JERRY_UNLIKELY (keys == NULL))
+    {
+      ecma_collection_free (return_names_p);
+      ecma_collection_free (visited_names_p);
+      return keys;
+    }
 #endif /* ENABLED (JERRY_ESNEXT) */
 
-    /* Second pass: collecting property names into an array. */
-    JMEM_DEFINE_LOCAL_ARRAY (names_p, all_properties_count, ecma_string_t *);
-
-    ecma_string_t **string_names_p = names_p + array_index_named_properties_count;
-#if ENABLED (JERRY_ESNEXT)
-    ecma_string_t **symbol_names_p = string_names_p + string_named_properties_count;
-#endif /* ENABLED (JERRY_ESNEXT) */
+    for (uint32_t i = 0; i < keys->item_count; i++)
+    {
+      ecma_value_t prop_name = keys->buffer_p[i];
+      ecma_string_t *name_p = ecma_get_prop_name_from_value (prop_name);
 
-    uint32_t array_index_name_pos = 0;
-    uint32_t string_name_pos = string_named_properties_count;
-    uint32_t lazy_string_name_pos = 0;
 #if ENABLED (JERRY_ESNEXT)
-    uint32_t symbol_name_pos = symbol_named_properties_count;
+      if (ecma_prop_name_is_symbol (name_p))
+      {
+        continue;
+      }
 #endif /* ENABLED (JERRY_ESNEXT) */
 
-    buffer_p = prop_names_p->buffer_p;
+      ecma_property_descriptor_t prop_desc;
+      ecma_value_t get_desc = ecma_op_object_get_own_property_descriptor (obj_p, name_p, &prop_desc);
 
-    for (uint32_t i = 0; i < prop_names_p->item_count; i++)
-    {
-      ecma_string_t *name_p = ecma_get_prop_name_from_value (buffer_p[i]);
-      ecma_ref_ecma_string (name_p);
-
-      uint32_t index = ecma_string_get_array_index (name_p);
-
-      if (index != ECMA_STRING_NOT_ARRAY_INDEX)
+      if (ECMA_IS_VALUE_ERROR (get_desc))
       {
-        JERRY_ASSERT (array_index_name_pos < array_index_named_properties_count);
+        ecma_collection_free (keys);
+        ecma_collection_free (return_names_p);
+        ecma_collection_free (visited_names_p);
+        return NULL;
+      }
 
-        uint32_t insertion_pos = 0;
-        while (insertion_pos < array_index_name_pos
-               && index > ecma_string_get_array_index (names_p[insertion_pos]))
-        {
-          insertion_pos++;
-        }
+      if (ecma_is_value_true (get_desc))
+      {
+        bool is_enumerable = (prop_desc.flags & ECMA_PROP_IS_ENUMERABLE) != 0;
+        ecma_free_property_descriptor (&prop_desc);
 
-        if (insertion_pos == array_index_name_pos)
-        {
-          names_p[array_index_name_pos++] = name_p;
-        }
-        else
+        if (ecma_has_string_value_in_collection (visited_names_p, name_p)
```

Comment:
At the moment the `ecma_has_string_value_in_collection` is guarded by the `JERRY_BUILTIN_JSON` macro. Using this here would mean that the method's header and implementation would need to move a bit. This is the reason for one of the Travis CI failure.